File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/mediator/solution/consumer.cpp
// consumer.cpp

#include <iostream>
#include "consumer.h"
#include "middleman.h"
#include "producer.h"

consumer::consumer(const std::string & a_name, middleman * a_middleman)
{
	my_name = a_name;
	my_middleman = a_middleman;
}

void consumer::consume()
{
	producer * a_producer = my_middleman->choose_producer();
	if (a_producer == nullptr)
	{
		throw dp_exception("void consumer::consume() -- a_producer == nullptr.");
	}
	
	a_producer->produce();
	std::cout << my_name << " consumes." << std::endl;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/mediator/solution/Consumer.java
// Consumer.java

public class Consumer
{
	public Consumer(String aName, Middleman aMiddleman)
	{
		myName = aName;
		myMiddleman = aMiddleman;
	}

	public void consume()
	{
		Producer aProducer = myMiddleman.chooseProducer();
		if (aProducer == null)
		{
			throw new DpException("void consume() -- aProducer == null.");
		}
		
		aProducer.produce();
		System.out.println(myName + " consumes.");
	}

	private String myName;
	private Middleman myMiddleman;
}
